File: toolwindow/src/main/kotlin/com/claudecodeplus/ui/viewmodels/tool/ToolTypes.kt
package com.claudecodeplus.ui.viewmodels.tool

/**
 * UI 层工具类型枚举
 *
 * 独立于 SDK，避免 UI 层直接依赖 SDK 枚举类型。
 * 这样可以在 SDK 变化时，只需修改 Mapper 层，UI 层保持不变。
 *
 * 支持的工具类型：
 * - 文件操作：READ, EDIT, MULTI_EDIT, WRITE, NOTEBOOK_EDIT
 * - 搜索类：GLOB, GREP
 * - 命令执行：BASH, BASH_OUTPUT, KILL_SHELL
 * - 任务管理：TASK, TODO_WRITE
 * - 网络类：WEB_FETCH, WEB_SEARCH
 * - MCP：MCP, LIST_MCP_RESOURCES, READ_MCP_RESOURCE
 * - 其他：EXIT_PLAN_MODE, SLASH_COMMAND, UNKNOWN
 */
enum class UiToolType {
    /** Bash 命令执行 */
    BASH,

    /** 文件编辑（单次替换） */
    EDIT,

    /** 文件批量编辑 */
    MULTI_EDIT,

    /** 文件读取 */
    READ,

    /** 文件写入 */
    WRITE,

    /** 文件模式匹配（glob pattern） */
    GLOB,

    /** 内容搜索（grep） */
    GREP,

    /** 待办事项管理 */
    TODO_WRITE,

    /** 子任务代理 */
    TASK,

    /** 网页抓取 */
    WEB_FETCH,

    /** 网络搜索 */
    WEB_SEARCH,

    /** Jupyter Notebook 编辑 */
    NOTEBOOK_EDIT,

    /** MCP 工具调用 */
    MCP,

    /** 获取后台 Bash 输出 */
    BASH_OUTPUT,

    /** 终止后台 Shell 进程 */
    KILL_SHELL,

    /** 退出计划模式 */
    EXIT_PLAN_MODE,

    /** 列出 MCP 资源 */
    LIST_MCP_RESOURCES,

    /** 读取 MCP 资源 */
    READ_MCP_RESOURCE,

    /** 自定义斜杠命令 */
    SLASH_COMMAND,

    /** 未知类型（兜底） */
    UNKNOWN
}

/**
 * IDE 集成类型
 *
 * 定义工具调用在 IDE 中的交互方式：
 * - SHOW_DIFF: 显示文件修改前后对比（用于 EDIT, MULTI_EDIT）
 * - OPEN_FILE: 在编辑器中打开文件（用于 READ, WRITE）
 */
enum class IdeIntegrationType {
    /** 在 IDE 中显示 Diff 对比视图 */
    SHOW_DIFF,

    /** 在 IDE 编辑器中打开文件 */
    OPEN_FILE
}

File: frontend/src/constants/toolTypes.ts
/**
 * 工具类型常量定义
 *
 * 命名规范:
 * - Claude SDK 工具: CLAUDE_<TOOL> (如 CLAUDE_READ, CLAUDE_EDIT)
 * - MCP 工具: MCP
 * - 未知工具: UNKNOWN
 */

/**
 * Claude SDK 工具类型常量
 */
export const CLAUDE_TOOL_TYPE = {
  READ: 'CLAUDE_READ',
  WRITE: 'CLAUDE_WRITE',
  EDIT: 'CLAUDE_EDIT',
  MULTI_EDIT: 'CLAUDE_MULTI_EDIT',
  BASH: 'CLAUDE_BASH',
  BASH_OUTPUT: 'CLAUDE_BASH_OUTPUT',
  KILL_SHELL: 'CLAUDE_KILL_SHELL',
  GREP: 'CLAUDE_GREP',
  GLOB: 'CLAUDE_GLOB',
  TODO_WRITE: 'CLAUDE_TODO_WRITE',
  WEB_SEARCH: 'CLAUDE_WEB_SEARCH',
  WEB_FETCH: 'CLAUDE_WEB_FETCH',
  TASK: 'CLAUDE_TASK',
  ASK_USER_QUESTION: 'CLAUDE_ASK_USER_QUESTION',
  NOTEBOOK_EDIT: 'CLAUDE_NOTEBOOK_EDIT',
  EXIT_PLAN_MODE: 'CLAUDE_EXIT_PLAN_MODE',
  ENTER_PLAN_MODE: 'CLAUDE_ENTER_PLAN_MODE',
  SKILL: 'CLAUDE_SKILL',
  SLASH_COMMAND: 'CLAUDE_SLASH_COMMAND',
  LIST_MCP_RESOURCES: 'CLAUDE_LIST_MCP_RESOURCES',
  READ_MCP_RESOURCE: 'CLAUDE_READ_MCP_RESOURCE',
} as const

/**
 * 其他工具类型常量
 */
export const OTHER_TOOL_TYPE = {
  MCP: 'MCP',
  UNKNOWN: 'UNKNOWN',
} as const

/**
 * Claude 工具类型联合类型
 */
export type ClaudeToolType = typeof CLAUDE_TOOL_TYPE[keyof typeof CLAUDE_TOOL_TYPE]

/**
 * 其他工具类型联合类型
 */
export type OtherToolType = typeof OTHER_TOOL_TYPE[keyof typeof OTHER_TOOL_TYPE]

/**
 * 所有工具类型联合类型
 */
export type ToolType = ClaudeToolType | OtherToolType

/**
 * 工具名称到类型的映射（用于前端本地转换，当后端未提供 toolType 时使用）
 */
export const TOOL_NAME_TO_TYPE: Record<string, ToolType> = {
  'Read': CLAUDE_TOOL_TYPE.READ,
  'Write': CLAUDE_TOOL_TYPE.WRITE,
  'Edit': CLAUDE_TOOL_TYPE.EDIT,
  'MultiEdit': CLAUDE_TOOL_TYPE.MULTI_EDIT,
  'Bash': CLAUDE_TOOL_TYPE.BASH,
  'BashOutput': CLAUDE_TOOL_TYPE.BASH_OUTPUT,
  'KillShell': CLAUDE_TOOL_TYPE.KILL_SHELL,
  'Grep': CLAUDE_TOOL_TYPE.GREP,
  'Glob': CLAUDE_TOOL_TYPE.GLOB,
  'TodoWrite': CLAUDE_TOOL_TYPE.TODO_WRITE,
  'WebSearch': CLAUDE_TOOL_TYPE.WEB_SEARCH,
  'WebFetch': CLAUDE_TOOL_TYPE.WEB_FETCH,
  'Task': CLAUDE_TOOL_TYPE.TASK,
  'AskUserQuestion': CLAUDE_TOOL_TYPE.ASK_USER_QUESTION,
  'NotebookEdit': CLAUDE_TOOL_TYPE.NOTEBOOK_EDIT,
  'ExitPlanMode': CLAUDE_TOOL_TYPE.EXIT_PLAN_MODE,
  'EnterPlanMode': CLAUDE_TOOL_TYPE.ENTER_PLAN_MODE,
  'Skill': CLAUDE_TOOL_TYPE.SKILL,
  'SlashCommand': CLAUDE_TOOL_TYPE.SLASH_COMMAND,
  'ListMcpResourcesTool': CLAUDE_TOOL_TYPE.LIST_MCP_RESOURCES,
  'ReadMcpResourceTool': CLAUDE_TOOL_TYPE.READ_MCP_RESOURCE,
}

/**
 * 辅助函数：根据工具名称解析类型
 *
 * @param name 工具名称（如 "Read", "mcp__excel__read"）
 * @returns 工具类型标识（如 "CLAUDE_READ", "MCP"）
 */
export function resolveToolType(name: string): ToolType {
  if (name.startsWith('mcp__')) {
    return OTHER_TOOL_TYPE.MCP
  }
  return TOOL_NAME_TO_TYPE[name] || OTHER_TOOL_TYPE.UNKNOWN
}
